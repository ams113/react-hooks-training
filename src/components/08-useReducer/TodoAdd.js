import React from 'react';
import { useForm } from '../../hooks/useForm';
import PropTypes from 'prop-types';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        const desc = description.trim();

        if ( desc.length <= 1) {
            return;
        }
        
        const newTodo = {
            id: new Date().getTime(),
            desc,
            done: false,
        };

        handleAddTodo( newTodo );

        reset();

    };

    return (
        <>
            <h4>Add TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    className="form-control"
                    name="description"
                    placeholder="Learn ..."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />
                <div className="d-grid gap-2">
                    <button
                        type="submit"
                        className="btn btn-outline-primary btn-block">
                            Add
                    </button>
                </div>
            </form>
        </>
    )
}

TodoAdd.propTypes = {
    handleAddTodo: PropTypes.func.isRequired,
};