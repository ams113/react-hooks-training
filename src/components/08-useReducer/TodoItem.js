import React from 'react';
import PropTypes from 'prop-types';

export const TodoItem = ({ item, index, handleDelete, handleToggle }) => {
    
    return (

        <li key={item.id} className="list-group-item">
            <p
                className={`${item.done && 'completed'}`} 
                onClick={ () => handleToggle(item.id) }
            >
                {index + 1}. {item.desc}
            </p>

            <button
                className="btn btn-danger"
                onClick={() => handleDelete(item.id)}
            >
                Delete
            </button>
        </li>
    )
}

TodoItem.propTypes = {
    item: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
};