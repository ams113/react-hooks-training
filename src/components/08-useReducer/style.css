.ml-5 {
    margin-left: 5px;
}
.mr-15 {
    margin-right: 15px;
}
.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
}

.completed{
    text-decoration: line-through;
}

p {
    cursor: pointer;
    margin:  0px;
}