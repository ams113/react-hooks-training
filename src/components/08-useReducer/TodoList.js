import React from 'react'
import PropTypes from 'prop-types';
import { TodoItem } from './TodoItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {

    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((item, index) => (
                    <TodoItem 
                        key={item.id}
                        item={item}
                        index={index}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.arrayOf(PropTypes.object).isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
};