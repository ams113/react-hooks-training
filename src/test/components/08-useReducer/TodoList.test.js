import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { fakeTodos } from '../../fixtures/fakeTodos';
import { TodoList } from '../../../components/08-useReducer/TodoList';

describe('Testing Components: TodoList', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList 
            todos={ fakeTodos } 
            handleDelete={handleDelete} 
            handleToggle={handleToggle}
        />
    );

    test('should renders correctly', () => {

        expect( wrapper ).toMatchSnapshot();
    });

    test('should have two <TodoItem />', () => {
        
        expect( wrapper.find('TodoItem').length ).toBe( fakeTodos.length );
        expect( wrapper.find('TodoItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
    });
    
});
