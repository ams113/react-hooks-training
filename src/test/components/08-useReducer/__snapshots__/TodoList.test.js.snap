// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Components: TodoList should renders correctly 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    item={
      Object {
        "desc": "Learn React",
        "done": false,
        "id": 1,
      }
    }
    key="1"
  />
  <TodoItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    item={
      Object {
        "desc": "Learn CSS",
        "done": false,
        "id": 2,
      }
    }
    key="2"
  />
</ul>
`;
